### PR TITLE
feat: pair-only chainlink oracle

### PR DESCRIPTION
**Pull request by totomanov**

### CODE REVIEW

File: src/chainlink/ChainlinkOracle.sol [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        } else {
+            return (baseAmount * NUMERATOR_SCALAR * FEED_SCALAR) / (answer * DENOMINATOR_SCALAR);
+        }
+    }
+
+    /// @inheritdoc IOracle
+    function priceOf(address base, address quote) external view returns (uint256) {
+        bool isForward = _getQueryDirection(base, quote);
+        uint256 answer = _getAnswer(base, quote);
+
```
Comment (alcueca) [inline src/chainlink/ChainlinkOracle.sol:49]: This returns prices as the value of one asset unit in terms of the other, while the spec states the prices should always be FP18.

File: src/chainlink/ChainlinkOracle.sol [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        if (isForward) {
+            return answer * DENOMINATOR_SCALAR / FEED_SCALAR;
```
Comment (alcueca) [inline src/chainlink/ChainlinkOracle.sol]:
```suggestion
            return answer * 1e18 / FEED_SCALAR;
```

File: src/chainlink/ChainlinkOracle.sol [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        } else {
+            return NUMERATOR_SCALAR * FEED_SCALAR / answer;
```
Comment (alcueca) [inline src/chainlink/ChainlinkOracle.sol]:
```suggestion
            return 1e18 * FEED_SCALAR / answer;
```

File: src/chainlink/ChainlinkOracle.sol [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        }
+    }
+
+    function _getAnswer(address base, address quote) internal view returns (uint256) {
```
Comment (alcueca) [inline src/chainlink/ChainlinkOracle.sol:62]: I like natspec in internal functions as well.

**Review by alcueca — CHANGES_REQUESTED** [review]
I wonder if it is too confusing and unexpected that prices are always FP18 regardless of the quote asset. Everyone seems to get it wrong.

**Comment by totomanov** [comment]
> I wonder if it is too confusing and unexpected that prices are always FP18 regardless of the quote asset. Everyone seems to get it wrong.

I do find it sort of unintuitive. The interface currently mimics a DEX swap interface. `baseAmount` is in units of `base`, so why isn't the returned amount in units of `quote`?

Most applications that need oracles use them to arrive at a price-equivalent amount of something. In this case FP18 isn't helpful and the consumer will need to convert decimals after getting the value.

**Comment by alcueca** [comment]
> why isn't the returned amount in units of quote?

Because you use `valueOf` for that.

> Most applications that need oracles use them to arrive at a price-equivalent amount of something. In this case FP18 isn't helpful and the consumer will need to convert decimals after getting the value.

True, but to get the "price-equivalent amount of something" you use `valueOf`.

`priceOf` doesn't return a price-equivalent amount of something, it returns a factor that ignores that the assets might have different decimals, like this:

image
[link](https://www.coinbase.com/en-pt/converter/eth/usdc)

**Review by alcueca — APPROVED** [review]
LGTM, but let's discuss whether `priceOf` shoould return an FP18, or be like `valueOf(10**base.decimals())`.

**Pull request merged**